hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/utils/CommitMetadataUtils.java

```diff
@@ -0,0 +1,220 @@
+    if (!table.getMetaClient().getTableConfig().getTableType().equals(HoodieTableType.MERGE_ON_READ)
+        || !commitActionType.equals(HoodieActiveTimeline.DELTA_COMMIT_ACTION)) {
+      return commitMetadata;
+    }
+
+    WriteMarkers markers = WriteMarkersFactory.get(config.getMarkersType(), table, instantTime);
+    // if there is log files in this delta commit, we search any invalid log files generated by failed spark task
+    boolean hasLogFileInDeltaCommit = commitMetadata.getPartitionToWriteStats()
+        .values().stream().flatMap(List::stream)
+        .anyMatch(writeStat -> FSUtils.isLogFile(new Path(config.getBasePath(), writeStat.getPath()).getName()));
+    if (hasLogFileInDeltaCommit) { // skip for COW table
+      // get all log files generated by makers
+      Set<String> allLogFilesMarkerPath = new HashSet<>(markers.getAppendedLogPaths(context, config.getFinalizeWriteParallelism()));
+      Set<String> logFilesMarkerPath = new HashSet<>();
+      allLogFilesMarkerPath.stream().filter(logFilePath -> !logFilePath.endsWith("cdc")).forEach(logFilePath -> logFilesMarkerPath.add(logFilePath));
+
+      // remove valid log files
+      for (Map.Entry<String, List<HoodieWriteStat>> partitionAndWriteStats : commitMetadata.getPartitionToWriteStats().entrySet()) {
+        for (HoodieWriteStat hoodieWriteStat : partitionAndWriteStats.getValue()) {
+          logFilesMarkerPath.remove(hoodieWriteStat.getPath());
+        }
+      }
+
+      // remaining are log files generated by retried spark task, let's generate write stat for them
+      if (logFilesMarkerPath.size() > 0) {
+        SerializableConfiguration serializableConfiguration = new SerializableConfiguration(hadoopConf);
+        context.setJobStatus(classNameForContext, "Preparing data for missing files to assit with generatin write stats");
+        // populate partition -> map (fileId -> HoodieWriteStat) // we just need one write stat per fileID to fetch some info about
+        // the file slice of interest to populate WriteStat.
+        HoodiePairData<String, Map<String, HoodieWriteStat>> partitionToWriteStatHoodieData = getPartitionToFileIdToFilesMap(commitMetadata, context);
+
+        String localBasePath = config.getBasePath();
+        // populate partition -> map (fileId -> List <missing log file names>)
+        HoodiePairData<String, Map<String, List<String>>> partitionToMissingLogFilesHoodieData =
+            getPartitionToFileIdToMissingLogFileMap(localBasePath, logFilesMarkerPath, context, config.getFileListingParallelism());
+
+        // TODO: make one call per partition to fetch file sizes.
+        context.setJobStatus(classNameForContext, "Generating writeStat for missing log files");
+
+        // lets join both to generate write stats for missing log files
+        List<Pair<String, List<HoodieWriteStat>>> additionalLogFileWriteStat = partitionToWriteStatHoodieData
+            .join(partitionToMissingLogFilesHoodieData)
```

Comment:
if there is no data file produced i.e. if a file slice is not part of in memory WriteStatus, I don't think there could be a scenario where we find additional missing files. These missing files are created due to spark retries. 
and this is targetting a successful commit. So, I can't think of a scenario where in-memory write status does not have a file slice, while retries created them.

for a partially failed commit, it could happen which our markers will handle it.
but good point though.